Agent 6 GUI

ISauve opened the pull request:
<!--
Please remember to review the Datadog Contribution Guidelines https://github.com/DataDog/datadog-agent/blob/master/CONTRIBUTING.md if you have not yet done so.
-->

### What does this PR do?

Adds a browser GUI for controlling agent 6. 

Other than the files added to `cmd/gui/`, there was also a couple other changes made to the agent:
1. Adding `restart` (mac/linux) & `restart-service` (windows) commands to the CLI
2. Fixed a small bug in `pkg/collector/autodiscovery/autoconfig.go`: getAllConfigs() was using `ac.collect(pd)` instead of `pd.provider.Collect()`, causing it to only return *new* config files (relative to the last time `Collect()` had been called) instead of *all* config files
3. Added functions to `pkg/collector/collector.go` in order to support reloading all instances of a check by name
4. Added error logging in the `autodiscovery` package for when a check configuration file is incorrect, that way if a user tries to add a new config file and it isn't correctly written the error can be fetched and displayed (previously, the error messages were just being logged to agent.log - now they are collected and saved in the `"autoconfig/ConfigErrors"` expvar map)
5. The Go check loader (`pkg/collector/corechecks/loader.go`) wasn’t returning errors that it was catching (similar to point 4, they were just being logged to agent.log but not collected). Changed this so that if even 1 instance of a check fails, it isn't going to return a nil error. Modified autodiscovery’s `GetChecks` method so that it registers these errors to the `”autoconfig/LoaderErrors”` expvar map. 

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

ISauve committed:
[gui] Initialized server, set authentication using JWT

ISauve committed:
Merge branch 'master' into isabelle/gui

ISauve committed:
[gui] Changed authentication to using API key, set up front end

ISauve committed:
[gui] Built the front end, integrated into the agent

ISauve committed:
[gui] Fixed communication error

ISauve committed:
[gui] Added some key functionalities

ISauve committed:
Merge branch 'master' into isabelle/gui

ISauve committed:
[gui] Cleaning up

ISauve committed:
[gui] Formatted status output

ISauve committed:
[gui] Added authentication secured pages, worked more on status templates, refactored

ISauve committed:
Merge branch 'master' into isabelle/gui

ISauve committed:
[gui] Added logs

ISauve committed:
[gui] Finished flare

ISauve committed:
[gui] Small refactoring, added fetch list of running checks

ISauve committed:
Merge branch 'master' into isabelle/gui

nmuesch committed:
API Keys can be 32 characters in lenght, so this increases the max char length

ISauve committed:
[gui] Bug fixes & other minor improvements

ISauve commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Experimental GUI](https://trello.com/c/aOTqsCDE/364-experimental-gui)

ISauve committed:
[gui] Added 'Add a Check' functionality

derekwbrown commented on cmd/gui/gui.go:
```diff
@@ -0,0 +1,154 @@
+	router := mux.NewRouter()
+
+	// Serve the (secured) index page on the default endpoint
+	router.Handle("/", accessAuth(http.FileServer(http.Dir("view/private/"))))
```
For  packaging, we shouldn't assume that the various support files are at a constant, relative path.  Make this a variable that we can change (at least at compile time) on a platform-specific basis.  This will make packaging much easier.

derekwbrown commented on cmd/gui/gui.go:
```diff
@@ -0,0 +1,154 @@
+	// Handle requests from clients
+	router.Handle("/req", authenticate(http.HandlerFunc(handler))).Methods("POST")
+
+	listener, e := net.Listen("tcp", port)
```
The server should definitely listen only on localhost `127.0.0.1`, not all interfaces (`0.0.0.0`)

ISauve replied:
It needs to run on 0.0.0.0 in case the user is running the agent on a VM or in a container. The fact that the interface is open to the whole local network was the original reason behind adding API key authentication (whether or not that is actually sufficient is another question though)

derekwbrown replied:
So, I'll approve as is so we can get this merged in.  But this is going to need to be re-thought.  API key auth isn't sufficient (for a lot of reasons), and open to everyone isn't going to fly.

ISauve committed:
[gui] Finished 'Add a Check' formatting, added ability to reload running checks

ISauve committed:
[gui] Check management

ISauve committed:
Merge branch 'master' into isabelle/gui

Conflicts:
	cmd/agent/app/start.go

ISauve committed:
[gui] Changed port host

ISauve committed:
[gui] Variable support file location

ISauve committed:
Merge branch 'master' into isabelle/gui

ISauve committed:
[gui] Documentation

ISauve committed:
[gui] Restart added: linux/mac

ISauve committed:
[gui] Flares allowed, windows restart added

ISauve committed:
[gui] Added 2 collector tests

ISauve committed:
Merge branch 'master' into isabelle/gui

ISauve committed:
[gui] Small bug fix (reload agent)

ISauve committed:
[gui] Changed windows restart agent

derekwbrown committed:
[Windows] Set up build for including windows GUI support files

ISauve committed:
Merge branch 'db/isabelle/gui' into isabelle/gui

ISauve committed:
[gui] Fixed path for support files (unix)

ISauve committed:
Merge branch 'master' into isabelle/gui

ISauve committed:
[gui] Refactored the GUI API, removed extra log files, added default check management

ISauve committed:
[gui] Refactored the GUI API, removed extra log files, added default check management

ISauve committed:
[gui] Error logging for incorrect check configuration files

ISauve committed:
Merge branch 'isabelle/gui' of github.com:DataDog/datadog-agent into isabelle/gui

Conflicts:
	cmd/gui/view/private/javascript.js

ISauve committed:
[gui] Added more error logging for Go check loader

ISauve committed:
Merge branch 'master' into isabelle/gui

Conflicts:
	cmd/agent/common/common_windows.go

ISauve committed:
[gui] Small refactor

ISauve committed:
[gui] Added check instance statuses to 'Checks Summary'

ISauve committed:
Merge branch 'master' into isabelle/gui

masci reviewed (changes requested):
Before we start reviewing throughly, a nit that has a big impact: can we move the `gui` folder within `cmd/agent`? I'd also rename `view` to `views`.

masci commented on cmd/gui/gui.go:
```diff
@@ -0,0 +1,137 @@
+	router.PathPrefix("/agent").Handler(negroni.New(negroni.HandlerFunc(authorize), negroni.Wrap(agentRouter)))
+	router.PathPrefix("/checks").Handler(negroni.New(negroni.HandlerFunc(authorize), negroni.Wrap(checkRouter)))
+
+	listener, e := net.Listen("tcp", ":"+port)
```
As @derekwbrown already pointed out, we cannot really ship the GUI like this. I see the value of accessing it from different networks but at the bare minimum make this configurable, defaulting to `localhost`.

derekwbrown replied:
For the moment, IMO, don't even make it configurable, only `localhost`.  Otherwise we open a can of worms we don't want to open WRT supporting remote administration via this gui.

derekwbrown reviewed (changes requested):
I think it looks good with the pair of changes outlined.

ISauve committed:
[gui] YAML syntax highlighting

ISauve committed:
[gui] Renamed view folder -> views, moved gui to cmd/agent

ISauve committed:
[gui] GUI fail doesn't break agent starting, limited interface

Removed ability for the GUI not starting to prevent start of base agent, limited network to local-interface only

ISauve committed:
[gui] Cleaning up expvar map

Stopping a check now removes that check from the runner["checks"] expvar map so that it won't show up under "Running Checks" when the collector status is rendered.
Correctly loading a configuration file for a check clears any previous entries for that check in the autoconfig["ConfigErrors"] expvar map.

derekwbrown committed:
Fix typo in function declaration GetViewsPath()

masci commented on cmd/agent/gui/views/public/js/yaml.js:
```diff
@@ -0,0 +1,118 @@
+// CodeMirror, copyright (c) by Marijn Haverbeke and others
```
we need to add this library to the file `LICENSE-3rdparty.csv`

ISauve replied:
Done :+1:

masci commented on pkg/collector/autodiscovery/autoconfig.go:
```diff
@@ -174,7 +176,19 @@ func (ac *AutoConfig) GetChecksByName(checkName string) []check.Check {
 func (ac *AutoConfig) getAllConfigs() []check.Config {
 	configs := []check.Config{}
 	for _, pd := range ac.providers {
-		cfgs, _ := ac.collect(pd)
```
I'm afraid this breaks Autodiscovery, `collect` does other things beside invoking `Collect()` on any provider

masci commented on pkg/collector/corechecks/loader.go:
```diff
@@ -56,15 +56,20 @@ func (gl *GoCheckLoader) Load(config check.Config) ([]check.Check, error) {
 	for _, instance := range config.Instances {
 		newCheck := factory()
 		if err := newCheck.Configure(instance, config.InitConfig); err != nil {
-			log.Errorf("core.loader: could not configure check %s: %s", newCheck, err)
```
should we keep the log line for those not using the GUI?

masci commented on pkg/collector/runner/runner.go:
```diff
@@ -279,6 +279,14 @@ func GetCheckStats() map[check.ID]*check.Stats {
 }
 
+// RemoveCheckFromMap removes a check from the check stats map
+func RemoveCheckFromMap(checkID check.ID) {
```
can we rename this to `RemoveCheckStats()`?

ISauve committed:
[gui] YAML syntax validation, changed restart

Added YAML sytanx validation, moved restart from common to gui package, changed linux/mac restart logic to no longer fork the process

ISauve committed:
Merge branch 'isabelle/gui' of github.com:DataDog/datadog-agent into isabelle/gui

ISauve commented on cmd/agent/gui/render.go:
```diff
@@ -0,0 +1,192 @@
+	return template.HTML(lastErrorArray[0]["traceback"])
+}
+
+func mkHuman(i int64) string {
```
One accepts floats, the other integers. I'll make that more clear

ISauve committed:
[gui] Small changes & bug fixes

Changed the GUI interface so that it's not responsible for not starting itself if the user de-activates it, renamed RemoveCheckFromMap -> RemoveCheckStats, added licence info for 3rd party packages, disabled restart for linux/mac, fixed firefox compatability issue

ISauve committed:
[gui] Added CodeMirror keymapping

YAML doesn't allow tabs, so this adds a keymapping to all the CodeMirror text editors which inserts spaces instead of a tab character

ISauve committed:
[gui] Linting issue & documentation

tmichelet committed:
[LogsAgent] Config - update configuration passed to the logs agent (#718)

We previously relied on `conf.d` being in the same directory as
`datadog.yaml`. This is the default setup but can be overriden.
Let's instead pass this information as two distinct parameters.

see https://github.com/DataDog/datadog-sheepdog-agent/pull/53

ISauve committed:
Merge branch 'master' into isabelle/gui

ISauve committed:
[gui] Disallow blank API keys

ISauve committed:
Changed invalidAPI.html message

ISauve committed:
[gui] Changed the 'add a check' interface

The text editor now automatically contains the example file's data for a check (if it has an example file)

ISauve committed:
Merge branch 'master' into isabelle/gui

masci committed:
Merge branch 'master' into isabelle/gui

ISauve committed:
[gui] Test error fix
